### PR TITLE
Defensive code for connection error handling and userId lookup.

### PR DESCRIPTION
**Pull request by SamTwining**
Defensive code first pass for connection error handling. Also sussed out a bug where it intercepted an invalid userId in the form of UserId|UserName (i.e. U82124|SamTwining) which choked the username lookup function.

Tested the connection error handling on a Win7 machine with no WebSocket support.

Handled the retry loop and then threw a descriptive error on failure:

image

### CODE REVIEW

File: lib/Slack.cs [hunk 1]
```diff
@@ -43,7 +43,7 @@ static class Slack
                     .Json
                     .JsonConvert
                     .DeserializeObject<SlackUserList>(responseContent)
-                    .Members.First(member => member.Id == userId).Name;
+                    .Members.First(member => member.Id == userId.Split('|')[0])?.Name ?? "";
```
Comment (SamTwining) [inline lib/Slack.cs]: Added a split for when it resolves the errant use case described in PR. Didn't want it to just pass back null, but there's a better way to handle only returning valid usernames.

File: lib/SocketConnection.cs [hunk 2]
```diff
@@ -11,49 +11,76 @@ class SocketConnection
...
+                            OnData.Invoke(this, rawMessage);
+                        }
+                    }
+                    break;
+                }
+                catch (Exception e)
                 {
-                    var receivedMessage = await Socket.ReceiveAsync(receiveBuffer, CancellationToken.None);
-                    if (receivedMessage.MessageType == WebSocketMessageType.Close)
+                    int sleepTimeSeconds = Convert.ToInt32(Math.Pow(secondsBetweenRetry, retryCounter + 1));
```
Comment (SamTwining) [inline lib/SocketConnection.cs]: Has an exponential backoff which felt pretty slow TBH, might want to consider tweaking with a linear time.

File: lib/SocketConnection.cs [hunk 3]
```diff
@@ -18,42 +18,63 @@ public SocketConnection(string url)
...
+                    retryCounter++;
+
+                    if (retryCounter >= maxRetryCount)
                     {
-                        await
-                            Socket.CloseAsync(WebSocketCloseStatus.NormalClosure, "closing websocket",
-                                CancellationToken.None);
+                        Console.WriteLine($"FATAL exception connecting to Slack: {Environment.NewLine} {e.ToString()}");
```
Comment (mattcbaker) [inline lib/SocketConnection.cs]: Attempting to not have super chatty console output

File: lib/SocketConnection.cs [hunk 4]
```diff
@@ -18,42 +18,63 @@ public SocketConnection(string url)
...
+                        throw e;
                     }
                     else
                     {
-                        var messageBytes = receiveBuffer.Skip(receiveBuffer.Offset).Take(receivedMessage.Count).ToArray();
-
-                        var rawMessage = new UTF8Encoding().GetString(messageBytes);
-                        OnData.Invoke(this, rawMessage);
+                        Thread.Sleep(sleepTimeSeconds * 1000);
+                        Connect();
```
Comment (mattcbaker) [inline lib/SocketConnection.cs]: Calling Connect after Thread.Sleep

File: lib/Slack.cs (hunk 1 above)
Comment (mattcbaker) [inline lib/Slack.cs]: I wonder if we could get this logic into SlackMessage.FindMentionedUsers. If we can, we can get it under test. So I can give it a whack, do you have the repro steps for this one?

**Pull request merged**